From ad3266250d0518031e97a7737c20ae8f744e2d50 Mon Sep 17 00:00:00 2001
From: Christoph Hellwig <hch@lst.de>
Date: Wed, 9 Sep 2020 17:31:33 +0900
Subject: [PATCH] elf2flt: add riscv 64-bits support

Add support for riscv 64bits ISA by defining the relocation types
R_RISCV_32_PCREL, R_RISCV_ADD32, R_RISCV_SUB32 and R_RISCV_64.
riscv64 support also needs a new linker script as the global_pointer
variable __global_pointer$ needs to be defined right after the
relocation tables in the data section and section ordering changes are
needed. With these changes, buildroot/busybox builds and run on NOMMU
systems with kernel 5.13. Tested on Canaan Kendryte K210 boards.

Signed-off-by: Christoph Hellwig <hch@lst.de>
Signed-off-by: Damien Le Moal <damien.lemoal@wdc.com>
---
 elf2flt.c     | 23 +++++++++++++++++++++++
 elf2flt.ld.in |  1 +
 ld-elf2flt.c  | 18 ++++++++++++++++++
 3 files changed, 42 insertions(+)

diff --git a/elf2flt.c b/elf2flt.c
index f87f1fc..981c325 100644
--- a/elf2flt.c
+++ b/elf2flt.c
@@ -80,6 +80,8 @@ const char *elf2flt_progname;
 #include <elf/v850.h>
 #elif defined(TARGET_xtensa)
 #include <elf/xtensa.h>
+#elif defined(TARGET_riscv64)
+#include <elf/riscv.h>
 #endif
 
 #if defined(__MINGW32__)
@@ -122,6 +124,8 @@ const char *elf2flt_progname;
 #define ARCH	"nios2"
 #elif defined(TARGET_xtensa)
 #define ARCH	"xtensa"
+#elif defined(TARGET_riscv64)
+#define ARCH	"riscv64"
 #else
 #error "Don't know how to support your CPU architecture??"
 #endif
@@ -797,6 +801,16 @@ output_relocs (
 					goto good_32bit_resolved_reloc;
 				default:
 					goto bad_resolved_reloc;
+#elif defined(TARGET_riscv64)
+				case R_RISCV_32_PCREL:
+				case R_RISCV_ADD32:
+				case R_RISCV_SUB32:
+					break;
+				case R_RISCV_32:
+				case R_RISCV_64:
+					goto good_32bit_resolved_reloc;
+				default:
+					goto bad_resolved_reloc;
 #else
 				default:
 					/* The default is to assume that the
@@ -1806,6 +1820,15 @@ int main(int argc, char *argv[])
   if (!load_to_ram && !pfile)
     load_to_ram = 1;
 
+#if defined(TARGET_riscv64)
+  /*
+   * riscv only supports loading text and data contiguously.
+   * So fail if load_to_ram is false.
+   */
+  if (!load_to_ram)
+    fatal("Loading to RAM ('-r' option) is required");
+#endif
+
   fname = argv[argc-1];
 
   if (pfile) {
diff --git a/elf2flt.ld.in b/elf2flt.ld.in
index ec1fe6f..b10ae84 100644
--- a/elf2flt.ld.in
+++ b/elf2flt.ld.in
@@ -70,6 +70,7 @@ W_RODAT		*(.gnu.linkonce.r*)
 		. = ALIGN(0x20) ;
 		LONG(-1)
 		. = ALIGN(0x20) ;
+RV64_GP		__global_pointer$ = . + 0x800;
 R_RODAT		*(.rodata)
 R_RODAT		*(.rodata1)
 R_RODAT		*(.rodata.*)
diff --git a/ld-elf2flt.c b/ld-elf2flt.c
index e5de506..113b8eb 100644
--- a/ld-elf2flt.c
+++ b/ld-elf2flt.c
@@ -324,6 +324,24 @@ static int do_final_link(void)
 		append_option(&other_options, concat(got_offset, "=", buf, NULL));
 	}
 
+	if (streq(TARGET_CPU, "riscv64")) {
+		/*
+		 * The .got section must come before the .got.plt section
+		 * (gcc/ld bug ?).
+		 */
+		append_sed(&sed, "(.got.plt)", "(.got.tmp)");
+		append_sed(&sed, "(.got.plt)", "(.got)");
+		append_sed(&sed, "(.got.tmp)", "(.got.plt)");
+
+		/*
+		 * The global pointer symbol is defined after the GOT.
+		 */
+		append_sed(&sed, "^RV64_GP", "");
+	} else {
+		/* Get rid of the global pointer definition */
+		append_sed(&sed, "^RV64_GP", NULL);
+	}
+
 	/* Locate the default linker script, if we don't have one provided. */
 	if (!linker_script)
 		linker_script = concat(ldscriptpath, "/elf2flt.ld", NULL);
-- 
2.31.1

